src/compiler/checker.ts

```diff
@@ -27715,6 +27737,27 @@ namespace ts {
             }
         }
 
+        function checkPrivateIdentifierExpression(privId: PrivateIdentifier): Type {
+            // The only valid position for a PrivateIdentifier to appear as an expression is on the left side of
+            // the `#field in expr` BinaryExpression
+            const isPrivateFieldInInExpression = isBinaryExpression(privId.parent)
+                && privId.parent.left === privId
+                && privId.parent.operatorToken.kind === SyntaxKind.InKeyword;
+            let symbolResolved = false;
+            if (isPrivateFieldInInExpression) {
+                const lexicallyScopedSymbol = lookupSymbolForPrivateIdentifierDeclaration(privId.escapedText, privId);
+                if (lexicallyScopedSymbol) {
+                    symbolResolved = true;
+                    markPropertyAsReferenced(lexicallyScopedSymbol, /* nodeForCheckWriteOnly: */ undefined, /* isThisAccess: */ false);
+                    getNodeLinks(privId.parent).resolvedSymbol = lexicallyScopedSymbol;
+                }
+            }
+            if (!symbolResolved) {
+                error(privId, Diagnostics.Private_identifiers_are_not_allowed_outside_class_bodies);

```

Comment:
This error is not 100% correct. We could be inside a class body though this does match the existing errors when handling invalid privateIds.

```ts
class C {
    f() {
        return #hello; // Private identifiers are not allowed outside class bodies.
    }
}
```

https://www.typescriptlang.org/play?#code/MYGwhgzhAEDC0G8BQ1XQGYAoCUiVoICcBTAFwFdCA7aAYgAtiQQB7Abn1QF8keg

Reply:
I've added that new error, and split the check into 3 levels.

1. are we in a class, otherwise there is no valid way to have a private identifier
2. is the privateIdentifier in a valid position
3. does the privateIdentifier match a private member